pkg/syncer/storage_policy_quota_reconciler.go

```diff
@@ -0,0 +1,248 @@
+limitations under the License.
+*/
+
+package syncer
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/client-go/util/workqueue"
+
+	cnsoperatorv1alpha1 "sigs.k8s.io/vsphere-csi-driver/v3/pkg/apis/cnsoperator"
+	storagepolicyusagev1alpha1 "sigs.k8s.io/vsphere-csi-driver/v3/pkg/apis/cnsoperator/storagepolicy/v1alpha1"
+	"sigs.k8s.io/vsphere-csi-driver/v3/pkg/csi/service/logger"
+	k8s "sigs.k8s.io/vsphere-csi-driver/v3/pkg/kubernetes"
+)
+
+// StoragePolicyQuotaReconciler is the interface for StoragePolicyQuota reconciler.
+type StoragePolicyQuotaReconciler interface {
+	// Run starts the reconciler.
+	Run(ctx context.Context, workers int)
+}
+
+type storagePolicyQuotaReconciler struct {
+	// metadataSyncer informer to get FSS information.
+	metadataSyncerInformer *metadataSyncInformer
+	// StoragePolicyQuota queue to add StoragePolicyUsage CRs.
+	svcQuotaAddQueue workqueue.RateLimitingInterface
+	// StoragePolicyQuota queue to delete StoragePolicyUsage CRs.
+	svcQuotaDelQueue workqueue.RateLimitingInterface
+	// StoragePolicyQuota Synced.
+	svcQuotaSynced cache.InformerSynced
+}
+
+// newStoragePolicyQuotaReconciler returns a StoragePolicyQuotaReconciler.
+func newStoragePolicyQuotaReconciler(
+	ctx context.Context,
+	resyncPeriod time.Duration,
+	metadataSyncerInformer *metadataSyncInformer,
+	svcStoragePolicyQuotaRateLimiter workqueue.RateLimiter,
+	stopCh <-chan struct{}) (*storagePolicyQuotaReconciler, error) {
+
+	log := logger.GetLogger(ctx)
+	// Create an informer for StoragePolicyQuota instances.
+	k8sConfig, err := k8s.GetKubeConfig(ctx)
+	if err != nil {
+		return nil, logger.LogNewErrorf(log, "newStoragePolicyQuotaReconciler: failed to get kubeconfig with error: %v", err)
+	}
+	dynamicInformerFactory, err := k8s.NewDynamicInformerFactory(ctx, k8sConfig, metav1.NamespaceAll, true)
+	if err != nil {
+		log.Errorf("newStoragePolicyQuotaReconciler: could not retrieve dynamic informer factory. Error: %+v", err)
+		return nil, err
+	}
+	// Return informer from shared dynamic informer factory for input resource.
+	gvr := schema.GroupVersionResource{Group: cnsoperatorv1alpha1.GroupName, Version: cnsoperatorv1alpha1.Version,
+		Resource: cnsoperatorv1alpha1.CnsStoragePolicyQuotaPlural}
+	storagePolicyQuotaInformer := dynamicInformerFactory.ForResource(gvr)
+	if err != nil {
+		return nil, logger.LogNewErrorf(log, "newStoragePolicyQuotaReconciler: failed to get informer with error: %v", err)
+	}
+	svcQuotaAddQueue := workqueue.NewNamedRateLimitingQueue(
+		svcStoragePolicyQuotaRateLimiter, "storage-policy-quota-add")
+	svcQuotaDelQueue := workqueue.NewNamedRateLimitingQueue(
+		svcStoragePolicyQuotaRateLimiter, "storage-policy-quota-del")
```

Comment:
Could you tell me the reason for having two separate queues? I feel it'll be racy, for example, if we have add SPQ, delete SPQ sequence, and it so happens that svcQuotaDelQueue processes it first, then the associated SPU may not get deleted.

Reply:
If there is single Queue, then you can differentiate between add and delete case using deletiontimestamp on the CR

Reply:
thanks for pointing it. For race conditions, incorrect ordering may cause issuer. will merge it into single queue.